From cc4d4ae5eb982eabce0a2296397d90e3d3e32727 Mon Sep 17 00:00:00 2001
From: "Srivatsa S. Bhat (VMware)" <srivatsa@csail.mit.edu>
Date: Thu, 16 Feb 2023 19:00:23 -0800
Subject: [PATCH] i40e: Fix build errors on 6.1.y

src/i40e_main.c:12357:17: error: too many arguments to function 'netif_napi_add'

Signed-off-by: Srivatsa S. Bhat (VMware) <srivatsa@csail.mit.edu>
---
 src/i40e_main.c        |  2 +-
 src/kcompat_impl.h     | 26 ++++++++++++++++++++++++++
 src/kcompat_std_defs.h | 13 +++++++++++++
 3 files changed, 40 insertions(+), 1 deletion(-)

diff --git a/src/i40e_main.c b/src/i40e_main.c
index 145ad7d..83b353e 100644
--- a/src/i40e_main.c
+++ b/src/i40e_main.c
@@ -12730,7 +12730,7 @@ static int i40e_vsi_alloc_q_vector(struct i40e_vsi *vsi, int v_idx)
 #endif
 	if (vsi->netdev)
 		netif_napi_add(vsi->netdev, &q_vector->napi,
-			       i40e_napi_poll, NAPI_POLL_WEIGHT);
+			       i40e_napi_poll);
 
 	/* tie q_vector and vsi together */
 	vsi->q_vectors[v_idx] = q_vector;
diff --git a/src/kcompat_impl.h b/src/kcompat_impl.h
index 4af4f98..724c0e8 100644
--- a/src/kcompat_impl.h
+++ b/src/kcompat_impl.h
@@ -846,4 +846,30 @@ void _kc_eth_hw_addr_set(struct net_device *dev, const void *addr);
 #endif /* eth_hw_addr_set */
 #endif /* NEED_ETH_HW_ADDR_SET */
 
+/*
+ * NEED_NETIF_NAPI_ADD_NO_WEIGHT
+ *
+ * Upstream commit b48b89f9c189 ("net: drop the weight argument from
+ * netif_napi_add") removes weight argument from function call.
+ *
+ * Our drivers always used default weight, which is 64.
+ *
+ * Define NEED_NETIF_NAPI_ADD_NO_WEIGHT on kernels 3.10+ to use old
+ * implementation. Undef for 6.1+ where new function was introduced.
+ */
+#ifdef NEED_NETIF_NAPI_ADD_NO_WEIGHT
+static inline void
+_kc_netif_napi_add(struct net_device *dev, struct napi_struct *napi,
+		   int (*poll)(struct napi_struct *, int))
+{
+	return netif_napi_add(dev, napi, poll, NAPI_POLL_WEIGHT);
+}
+
+/* RHEL7 complains about redefines. Undef first, then define compat wrapper */
+#ifdef netif_napi_add
+#undef netif_napi_add
+#endif
+#define netif_napi_add _kc_netif_napi_add
+#endif /* NEED_NETIF_NAPI_ADD_NO_WEIGHT */
+
 #endif /* _KCOMPAT_IMPL_H_ */
diff --git a/src/kcompat_std_defs.h b/src/kcompat_std_defs.h
index 361b72b..558b304 100644
--- a/src/kcompat_std_defs.h
+++ b/src/kcompat_std_defs.h
@@ -31,6 +31,12 @@
 #error "KERNEL_VERSION is undefined"
 #endif
 
+/*****************************************************************************/
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(3,10,0))
+#else /* >= 3,10,0 */
+#define NEED_NETIF_NAPI_ADD_NO_WEIGHT
+#endif /* 3,10,0 */
+
 /*****************************************************************************/
 #if (LINUX_VERSION_CODE < KERNEL_VERSION(3,18,0))
 #define NEED_DEV_PM_DOMAIN_ATTACH_DETACH
@@ -260,4 +266,11 @@
 #define HAVE_ETHTOOL_EXTENDED_RINGPARAMS
 #endif /* 5.17.0 */
 
+/*****************************************************************************/
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(6,1,0))
+#else /* >=6.1.0 */
+#undef NEED_NETIF_NAPI_ADD_NO_WEIGHT
+#undef HAVE_NDO_GET_DEVLINK_PORT
+#endif /* 6.1.0 */
+
 #endif /* _KCOMPAT_STD_DEFS_H_ */
-- 
2.25.1

